### ISSUE TITLE
If a wrist encoder is broken, the arms may move in a very weird configuration

### ISSUE THREAD

**Issue by S-Dafarra** [robotology/walking-teleoperation#52]
This happened once before a demo. We need to double-check if the IK is using the measured quantities instead of the desired ones.

cc @kouroshD @GiulioRomualdi

**Comment by kouroshD** [comment]
@GiulioRomualdi  @S-Dafarra  Guys, I made a calendar event for this issue on Wednesday afternoon to work on it. I book the robot for this purpose as well. Let me know if there is any problem with this event.

**Comment by GiulioRomualdi** [comment]
Regarding this task

The IK uses a kindyn object and retrieves all the data from there (joints and base)

https://github.com/robotology/walking-controllers/blob/dcc2698e3d16adb5c78c973a27a359f0986ae535/src/WalkingModule/src/Module.cpp#L385


before calling `solveQPIK` the kin dyn object is set with the desired data computed by the IK 
https://github.com/robotology/walking-controllers/blob/dcc2698e3d16adb5c78c973a27a359f0986ae535/src/WalkingModule/src/Module.cpp#L729 

So given my best understanding, the IK is not taking measured quantities. 

However, I noticed that the retargeting client is also reset using a kindyn object (this happens only at the end of `prepare`). So if the wrist is in a strange configuration (or the encoder gives nonsense data) the value of the wrist is used as a regularization term in the walking. This may go in contrast with the cartesian task.

https://github.com/robotology/walking-controllers/blob/dcc2698e3d16adb5c78c973a27a359f0986ae535/src/WalkingModule/src/Module.cpp#L507-L512

So in conclusion:
1. If the wrist goes in HF while the walking is running nothing should happen
2. if the encoder streams nonsense data the retargeting client is using the data as input and (in case of walking + oculus) is used as a regularization term. 

This analysis if valid only for walking + xsens and walking + oculus. 

To solve the problem we may add
```cpp
iDynTree::Vector zero(m_qDesired,size());
zero.zeros()
m_FKSolver->setInternalRobotState(m_qDesired, zero);
```
just before https://github.com/robotology/walking-controllers/blob/dcc2698e3d16adb5c78c973a27a359f0986ae535/src/WalkingModule/src/Module.cpp#L507-L512

**Comment by S-Dafarra** [comment]
> 2\. if the encoder streams nonsense data the retargeting client is using the data as input and (in case of walking + oculus) is used as a regularization term.

Indeed, this was happening only for the walking +  oculus, and when the wrist encoder was broken, outputting 500deg, so definitely nonsense.

**Comment by S-Dafarra** [comment]
I was thinking about it, and I guess it is not so simple. Having the measured quantities inside the FK solver makes absolute sense in that position. We need to initialize the filters, so it makes sense to start from the measured quantities. 

So, how can we distinguish between a good and a bad case?

Maybe adding some checks on the initial cartesian error?

**Comment by GiulioRomualdi** [comment]
or directly in the joints no?

**Comment by S-Dafarra** [comment]
The modifications proposed by @GiulioRomualdi in https://github.com/robotology/walking-teleoperation/issues/52#issuecomment-799584929 have been added with https://github.com/robotology/walking-controllers/pull/82/commits/9f1e3ac3755592e415898cf2be8b43b2b115815d on the PR https://github.com/robotology/walking-controllers/pull/82, and tested on the robot. That did not change the behavior on the robot, but should prevent other crazy behaviors when an encoder is not working properly.

**Comment by kouroshD** [comment]
We have also done some more fine-tuning for the hand retargeting in  https://github.com/robotology/walking-controllers/commit/150d6070b3f532914d6bfcd5bf3e0cca5391dfa5 and added the missing left wrist joints in the configuration files in https://github.com/robotology/walking-controllers/commit/f722f8cc20154e44f1991d2d5e0e843ddcb0c2a3 . 
While we were testing, we put also the robot wrist in idle to check if things are on, however this was not the replication of the problem.

**Comment by S-Dafarra** [comment]
SInce the PR is open, we can close this issue. We can open a new one if it happens again.
